pymc3/external/emcee/backends.py

```diff
@@ -0,0 +1,167 @@
+        if sampler_stats is not None:
+            for data, vars in zip(self._stats, sampler_stats):
+                for key, val in vars.items():
+                    data[key][self.draw_idx] = val
+        self.draw_idx += 1
+
+    def _get_sampler_stats(self, varname, sampler_idx, burn, thin):
+        return self._stats[sampler_idx][varname][burn::thin]
+
+    def close(self):
+        if self.draw_idx == self.draws:
+            return
+        # Remove trailing zeros if interrupted before completed all
+        # draws.
+        self.samples = {var: vtrace[:self.draw_idx]
+                        for var, vtrace in self.samples.items()}
+        if self._stats is not None:
+            self._stats = [{var: trace[:self.draw_idx] for var, trace in stats.items()}
+                           for stats in self._stats]
+
+    # Selection methods
+
+    def __len__(self):
+        if not self.samples:  # `setup` has not been called.
+            return 0
+        return self.draw_idx
+
+    def get_values(self, varname, burn=0, thin=1):
+        """Get values from trace.
+
+        Parameters
+        ----------
+        varname : str
+        burn : int
+        thin : int
+
+        Returns
+        -------
+        A NumPy array
+        """
+        return self.samples[varname][burn::thin]
+
+    def _slice(self, idx):
+        # Slicing directly instead of using _slice_as_ndarray to
+        # support stop value in slice (which is needed by
+        # iter_sample).
+
+        # Only the first `draw_idx` value are valid because of preallocation
+        idx = slice(*idx.indices(len(self)))
+
+        sliced = NDArray(model=self.model, vars=self.vars)
+        sliced.chain = self.chain
+        sliced.samples = {varname: values[idx]
+                          for varname, values in self.samples.items()}
+        sliced.sampler_vars = self.sampler_vars
+        sliced.draw_idx = (idx.stop - idx.start) // idx.step
+
+        if self._stats is None:
+            return sliced
+        sliced._stats = []
+        for vars in self._stats:
+            var_sliced = {}
+            sliced._stats.append(var_sliced)
+            for key, vals in vars.items():
+                var_sliced[key] = vals[idx]
+
+        return sliced
+
+    def point(self, idx):
+        """Return dictionary of point values at `idx` for current chain
+        with variable names as keys.
+        """
+        idx = int(idx)
+        return {varname: values[idx]
+                for varname, values in self.samples.items()}
+
+
+def repack_ensemble_trace(trace):
+    traces = []
+    for i in range(trace.nparticles):
+        tr = NDArray('mcmc', trace.model, trace.vars)
+        tr.setup(len(trace), i)
+        for varname in trace.varnames:
+            tr.samples[varname] = trace.samples[varname][:, i]
+            tr.draw_idx = len(trace)
+        traces.append(tr)
+    return MultiTrace(traces)
+
+
+def ensure_multitrace(trace):
```

Comment:
Trace can have `demendency` attribute `'pymc3'` or 'emcee' to make optional imports